feat: add E2B code interpreter

jamesmurdza opened the pull request:
This is a plug-and-play alternative of the Docker code interpreter that uses E2B cloud sandboxes.

The inputs and outputs are slightly different compared to the default code interpreter:

Input: Dependencies are specified in the code via "!pip install"
Output: Output is a JSON structure separating standard output, errors, and rich data.

However these don't necessarily affect the end user as they are specified for the agent in the schema.

It also requires an E2B API key.

jamesmurdza commented:
Example agent: https://gist.github.com/jamesmurdza/956021e18c9576560c01b25f3e9d5b6b

lorenzejay commented on crewai_tools/tools/e2b_code_interpreter_tool/README.md:
```diff
@@ -0,0 +1,35 @@
+from crewai import Agent
+from crewai_tools import E2BCodeInterpreterTool
+
+code_interpreter = E2BCodeInterpreterTool()
```
this is probably not needed if inserting directly onto tools list

lorenzejay commented on crewai_tools/tools/e2b_code_interpreter_tool/README.md:
```diff
@@ -0,0 +1,35 @@
+
+# ... Use the agent ...
+    
+code_interpreter.close()
```
should we just close when tool returns the result?

lorenzejay commented on crewai_tools/tools/e2b_code_interpreter_tool/code_interpreter_tool.py:
```diff
@@ -0,0 +1,60 @@
+        # Convert the result dictionary to a JSON string since CrewAI expects a string output
+        content = json.dumps(result, indent=2)
+        
+        return content
```
should we add `self.close()` then return content ?

jamesmurdza commented:
Hi @lorenzejay, I've just improved the text and added comments to the README to explain why we have a `close()` function. I also added arguments to the tool, including timeout.

lorenzejay commented:
> Hi @lorenzejay, I've just improved the text and added comments to the README to explain why we have a `close()` function. I also added arguments to the tool, including timeout.

Awesome. Again, should it just automatically use the close method once it finishes the response? Don't need to keep the process running when its done right. Instead of having users set close manually. not saying remove the method just include that or even include that as a flag users can set ie: `close_when_finished: bool`

jamesmurdza commented:
> > Hi @lorenzejay, I've just improved the text and added comments to the README to explain why we have a `close()` function. I also added arguments to the tool, including timeout.
> 
> Awesome. Again, should it just automatically use the close method once it finishes the response? Don't need to keep the process running when its done right. Instead of having users set close manually. not saying remove the method just include that or even include that as a flag users can set ie: `close_when_finished: bool`

@lorenzejay If I understand correctly that would close the sandbox after the first time the tool is called. Since we don't know how many times the tool will be called (and it's frequently used more than once) that would cause workflows to break.

lorenzejay commented:
@jamesmurdza  getting close please add `E2BCodeInterpreterTool` in __init__.py to properly import it from `crew_tools`

lorenzejay commented:
@jamesmurdza when running the tool on my crew i get this:

  File ".../pypoetry/virtualenvs/crewai-Nmm_nbE4-py3.11/lib/python3.11/site-packages/e2b_code_interpreter/main.py", line 40, in __init__
    super().__init__(
TypeError: Sandbox.__init__() got an unexpected keyword argument 'envs'

jamesmurdza commented:
@lorenzejay I'm seeing "1 requested change" but I think all the requested changes have been made. Is there anything else?

tizkovatereza commented:
Hello @lorenzejay , just pinging this PR
Thank you! :)

tonykipkemboi commented:
@jamesmurdza & @tizkovatereza are there any recent updates on E2B side that need to be added to this PR?

lucasgomide commented on crewai_tools/tools/e2b_code_interpreter_tool/code_interpreter_tool.py:
```diff
@@ -0,0 +1,68 @@
+
+    def _run(self, code: str) -> str:
+        # Execute the code using the code interpreter
+        execution = self._code_interpreter_tool.notebook.exec_cell(code)
```
would be great wrap this code with `try/catch`. Remember that tools **must** returns `string` whenever possible

lucasgomide replied:
use `finally` to ensure `close` connection also

lucasgomide reviewed (changes requested):
add some tests if possible
